### DISCUSSION TITLE
Could components be interfering with sizing?

### DISCUSSION THREAD

**Discussion by marklam** [fsprojects/Avalonia.FuncUI#349] [Q&A] [answered]
Sorry, I'm back with the ListView related questions :-(

If I put grids of a fixed height in a listview template, scrolling works fine. If I make components holding a grid, then dragging the scrollbars goes crazy.

I've got a [project](https://github.com/marklam/FuncUIFileDialog/commit/2698c06f6b808ab290389290db526dc12dbd852b) or [script](https://github.com/marklam/FuncUIFileDialog/blob/main/ListBoxOfComponent/ListBoxOfComponent.fsx) that demonstrates it.

In the debug inspector, the component looks to have the correct height, but the list view looks like it doesn't know how big the contents are (even with the hint that `AreVerticalSnapPointsRegular=true`)

Here's what scrolling the grids (left) and components (right) look like:

![ScrollListOfComponents2](https://github.com/fsprojects/Avalonia.FuncUI/assets/1962640/9e09b37e-6e9b-4319-b259-eb602278ef4b)

Again thanks for your time, and for the very cool library.

**Comment by JaggerJo** [comment]
Hey @marklam 

can you try changing `ItemsControl` for a `ListBox` just to see if performance improves?

**Comment by marklam** [comment]
The ItemsControl of Grids renders, but the ItemsControl of Components doesn't appear, I get the wait cursor and the call stack shows it's stuck here:
<img src="https://github.com/fsprojects/Avalonia.FuncUI/assets/1962640/4569beb0-2e20-4aeb-8a27-9caebf2a40c3" height="800px"/>
[Updated code](https://github.com/marklam/FuncUIFileDialog/commit/0203bdc4900b4affd66eb4d418d03017dcc7fb8f)

**Comment by marklam** [comment]
If the core of the Component's `Update` is called synchronously from `OnInitialized`, rather than being queued on the dispatcher, the problem goes away.

I *believe* that it's valid to do that, because OnInitialized will be called on the UI thread anyway (`Dispatcher.UIThread.CheckAccess()` returned true)

The problem with the code as it stands seems to be that after OnInitialized, the component isn't sized correctly, it only gets the size set when the queued action is executed, at which point the scroll bars jiggle etc.

[This changeset](https://github.com/fsprojects/Avalonia.FuncUI/commit/1f100d9fdb6c1ac9183b447d0757b4e894b4c9d5) shows my proposed fix. 
* I've pulled the Action out of `Update` into `UIThreadUpdate`. 
* `Update` queues a call to UIThreadUpdate instead of the nested Action
* `OnInitialized` calls direct to UIThreadUpdate, rather than queueing via `Update` (as it does in the current code)

I can create a PR if you think it's a valid approach.

**Reply by JaggerJo** [reply]
Seems like a good approach. Thanks for investigating 👍

**Reply by marklam** [reply, marked as answer]
Created a PR 
https://github.com/fsprojects/Avalonia.FuncUI/pull/352
